presto-spi/src/main/java/com/facebook/presto/spi/relation/LogicalRowExpressions.java

```diff
@@ -467,95 +475,272 @@ public RowExpression visitVariableReference(VariableReferenceExpression referenc
+            return new ConvertNormalFormVisitorContext(clauseJoiner, depth + 1);
         }
     }
 
     private class ConvertNormalFormVisitor
-            implements RowExpressionVisitor<RowExpression, Form>
+            implements RowExpressionVisitor<RowExpression, ConvertNormalFormVisitorContext>
     {
         @Override
-        public RowExpression visitSpecialForm(SpecialFormExpression specialForm, Form clauseJoiner)
+        public RowExpression visitSpecialForm(SpecialFormExpression specialFormExpression, ConvertNormalFormVisitorContext context)
         {
-            if (!isConjunctionOrDisjunction(specialForm)) {
-                return specialForm;
+            if (!isConjunctionOrDisjunction(specialFormExpression)) {
+                return specialFormExpression;
+            }
+            // Attempt to convert sub predicates to expected normal form, deduplicate and fold constants.
+            RowExpression rewritten = combinePredicates(
+                    specialFormExpression.getForm(),
+                    extractPredicates(specialFormExpression.getForm(), specialFormExpression)
+                            .stream()
+                            .map(subPredicate -> subPredicate.accept(this, context.childContext()))
+                            .collect(toList()));
+
+            if (!isConjunctionOrDisjunction(rewritten)) {
+                return rewritten;
+            }
+
+            SpecialFormExpression rewrittenSpecialForm = (SpecialFormExpression) rewritten;
+            Form currentForm = rewrittenSpecialForm.getForm();
+            List<List<RowExpression>> subPredicates = getSubPredicates(rewrittenSpecialForm);
+
+            if (subPredicates.stream().mapToInt(List::size).sum() > ELIMINATE_COMMON_SIZE_LIMIT) {
+                return combineSubPredicates(currentForm, subPredicates);
             }
+            subPredicates = eliminateCommonPredicates(subPredicates);
 
-            // normalize arguments
-            RowExpression left = specialForm.getArguments().get(0).accept(new ConvertNormalFormVisitor(), clauseJoiner);
-            RowExpression right = specialForm.getArguments().get(1).accept(new ConvertNormalFormVisitor(), clauseJoiner);
+            //extractCommonPredicates when currentForm is same as clauseJoiner can produce opposite form
+            List<List<RowExpression>> flippedSubPredicates = extractCommonPredicates(currentForm, subPredicates);
+            if (flippedSubPredicates != null) {
+                subPredicates = flippedSubPredicates;
+                currentForm = flip(currentForm);
+            }
+
+            int currentSize = subPredicates.stream().mapToInt(List::size).sum();
+
+            // The end result can be a cartesian cross product list of list
+            int expectedSize;
+            try {
+                expectedSize = multiplyExact(subPredicates.stream().mapToInt(List::size).reduce(1, Math::multiplyExact), subPredicates.size());
```

Comment:
As discussed offline, use  for-loop so we can return early when `expectedSize` is too large. 

Using exception to handle normal business logic is not a good practice and should be avoided.